llvm/test/Transforms/LoopVectorize/AArch64/simple_early_exit.ll

```diff
@@ -0,0 +1,2544 @@
+; CHECK-NEXT:    [[WIDE_LOAD:%.*]] = load <vscale x 16 x i8>, ptr [[TMP29]], align 1
+; CHECK-NEXT:    [[TMP30:%.*]] = getelementptr inbounds i8, ptr [[P2]], i64 [[TMP12]]
+; CHECK-NEXT:    [[TMP31:%.*]] = getelementptr inbounds i8, ptr [[TMP30]], i32 0
+; CHECK-NEXT:    [[WIDE_LOAD2:%.*]] = load <vscale x 16 x i8>, ptr [[TMP31]], align 1
+; CHECK-NEXT:    [[TMP32:%.*]] = icmp eq <vscale x 16 x i8> [[WIDE_LOAD]], [[WIDE_LOAD2]]
+; CHECK-NEXT:    [[TMP33:%.*]] = xor <vscale x 16 x i1> [[TMP32]], shufflevector (<vscale x 16 x i1> insertelement (<vscale x 16 x i1> poison, i1 true, i64 0), <vscale x 16 x i1> poison, <vscale x 16 x i32> zeroinitializer)
+; CHECK-NEXT:    [[TMP34:%.*]] = call i1 @llvm.vector.reduce.or.nxv16i1(<vscale x 16 x i1> [[TMP33]])
+; CHECK-NEXT:    br i1 [[TMP34]], label [[VECTOR_EARLY_EXIT:%.*]], label [[FOR_INC3]]
+; CHECK:       loop.inc3:
+; CHECK-NEXT:    [[INDEX_NEXT4]] = add nuw i64 [[INDEX1]], [[TMP7]]
+; CHECK-NEXT:    [[TMP35:%.*]] = icmp eq i64 [[INDEX_NEXT4]], [[N_VEC]]
+; CHECK-NEXT:    br i1 [[TMP35]], label [[MIDDLE_BLOCK:%.*]], label [[VECTOR_BODY]], !llvm.loop [[LOOP0:![0-9]+]]
+; CHECK:       vector.early.exit:
+; CHECK-NEXT:    [[TMP36:%.*]] = call i64 @llvm.experimental.cttz.elts.i64.nxv16i1(<vscale x 16 x i1> [[TMP33]], i1 true)
+; CHECK-NEXT:    [[TMP37:%.*]] = add i64 [[TMP36]], [[INDEX1]]
+; CHECK-NEXT:    [[IND_EARLY_ESCAPE:%.*]] = add i64 3, [[TMP37]]
+; CHECK-NEXT:    br label [[FOR_END_LOOPEXIT:%.*]]
+; CHECK:       middle.block:
+; CHECK-NEXT:    [[CMP_N:%.*]] = icmp eq i64 40, [[N_VEC]]
+; CHECK-NEXT:    br i1 [[CMP_N]], label [[FOR_END_LOOPEXIT]], label [[SCALAR_PH]]
+; CHECK:       scalar.ph:
+; CHECK-NEXT:    [[BC_RESUME_VAL:%.*]] = phi i64 [ [[IND_END]], [[MIDDLE_BLOCK]] ], [ 3, [[ENTRY:%.*]] ]
+; CHECK-NEXT:    br label [[LAND_RHS:%.*]]
+; CHECK:       loop:
+; CHECK-NEXT:    [[INDEX:%.*]] = phi i64 [ [[INDEX_NEXT:%.*]], [[FOR_INC:%.*]] ], [ [[BC_RESUME_VAL]], [[SCALAR_PH]] ]
+; CHECK-NEXT:    [[ARRAYIDX:%.*]] = getelementptr inbounds i8, ptr [[P1]], i64 [[INDEX]]
+; CHECK-NEXT:    [[TMP38:%.*]] = load i8, ptr [[ARRAYIDX]], align 1
+; CHECK-NEXT:    [[ARRAYIDX1:%.*]] = getelementptr inbounds i8, ptr [[P2]], i64 [[INDEX]]
+; CHECK-NEXT:    [[TMP39:%.*]] = load i8, ptr [[ARRAYIDX1]], align 1
+; CHECK-NEXT:    [[CMP3:%.*]] = icmp eq i8 [[TMP38]], [[TMP39]]
+; CHECK-NEXT:    br i1 [[CMP3]], label [[FOR_INC]], label [[FOR_END_LOOPEXIT]]
+; CHECK:       loop.inc:
+; CHECK-NEXT:    [[INDEX_NEXT]] = add i64 [[INDEX]], 1
+; CHECK-NEXT:    [[EXITCOND:%.*]] = icmp ne i64 [[INDEX_NEXT]], 43
+; CHECK-NEXT:    br i1 [[EXITCOND]], label [[LAND_RHS]], label [[FOR_END_LOOPEXIT]], !llvm.loop [[LOOP3:![0-9]+]]
+; CHECK:       loop.end:
+; CHECK-NEXT:    [[START_0_LCSSA:%.*]] = phi i64 [ [[INDEX]], [[LAND_RHS]] ], [ 43, [[FOR_INC]] ], [ [[IND_EARLY_ESCAPE]], [[VECTOR_EARLY_EXIT]] ], [ 43, [[MIDDLE_BLOCK]] ]
+; CHECK-NEXT:    ret i64 [[START_0_LCSSA]]
+;
```

Comment:
Hi @david-arm , we have already talked about this a little, but I was wondering about the profitability of vectorising these cases. This is an example of a loop with a very low trip count, and I cannot imagine that vectorisation of this particular case will be very beneficial. Do you think think we might need some heuristics and profitability checks? Do you have data and/or opinions on this?

Reply:
That's a good question! Well, the program `xz` is one such example where we already vectorise an early-exit loop using the AArch64LoopIdiomTransform pass, where the loop is known to have low trip counts. That pass uses a predicated loop as the most optimal solution, but even an unpredicated vector loop with a scalar epilogue was still seen as an improvement on the scalar loop. So it's not entirely obvious to me that we shouldn't be vectorising loops with low trip counts.

I think ultimately we should let the vectoriser decide the best approach based on the cost model, and there is already code in the vectoriser that takes into account the cost of runtime checks and tries to determine the minimum trip count (if known at compile time) required to be profitable. Perhaps as part of this work I could add the cost of the cttz intrinsic in a similar way so that it can make a more informed choice?

Reply:
I've added the cost of work required in the early exit block when assessing profitability of vectorising. This will help define the minimum profitable trip count required in order to enter the vector loop.

Reply:
I will independently create a separate patch that improves the cost model for the cttz.elts intrinsic. For now I've disabled early-exit autovectorisation until we've got a better cost model, but that shouldn't hold up this patch.

Reply:
Hi, sorry for the late reply. I wanted to elaborate a little bit about why I brought up cost-modeling. It's kind of all in the name with "simple early exit": I was wondering if vectorisation would be beneficial if loops exit early. We now have two code paths: the early exit and the full vector body, which makes the perfomance modeling and characteristics let's say less uniform. That's why I wrote a couple of micro benchmark to test this, i.e. I tried to find at what iteration taking the early exit would give a performance uplift. For simple loops, I found you'll need at least a couple of hundred iterations for this vectorisation strategy to be beneficial. This is where my first observation of low trip counts came from: when we execute just a couple of iterations, the overhead of vectorisation is not worth it and actually results in (minor) regressions compared to scalar code. 

I hope I explained this a little bit better now. And by the way, there's probably a lot more going on than just looking at low trip count, but again this was just a first observation.

When the trip count isn't known at compile time, there's nothing we can do, I think. But when the trip count is a small constant, like some of the regression test cases, that looks like something we could handle.

I haven't looked at your latest changes related to this, just wanted to leave this comment first, but will take a look.